pyccel/parser/semantic.py

```diff
@@ -1108,11 +1143,37 @@ def _assign_lhs_variable(self, lhs, d_var, rhs, new_expressions, is_augassign,ar
             if not arr_in_multirets:
                 self._ensure_target(rhs, d_lhs)
 
-            var = self.check_for_variable(name)
-
+            if isinstance(lhs, DottedName):
+                var = self.check_for_attribute(lhs)
+            else:
+                var = self.check_for_variable(name)
             # Variable not yet declared (hence array not yet allocated)
             if var is None:
-
+                if isinstance(lhs, DottedName):
+                    if (self._current_function == '__init__' or self.current_class) and lhs.name[0] == 'self':
+                        var      = self.get_variable('self')
+                        cls_name = str(var.cls_base.name)
+                        cls      = self.scope.find(cls_name, 'classes')
+
+                        attributes = cls.attributes
+                        parent     = cls.superclasses
+                        attributes = list(attributes)
+                        name     = str(lhs.name[-1])
+
+                        member = self._create_variable(name, dtype, rhs, d_lhs)
+                        lhs    = member.clone(member.name, new_class = DottedVariable, lhs = var)
+
+                        # update the attributes of the class and push it to the scope
+                        attributes += [member]
+                        if not self._current_class:
+                            new_cls = ClassDef(cls_name, attributes, [], superclasses=parent)
+                            self._current_class = new_cls
```

Comment:
Is this possible? The coverage bot will make you test it, but I am not clear on what case it covers

Reply:
It will be tested each time their is a new creating of a class (precisely in `__init__` method).

Reply:
it is, but the `self.current_class` remain empty.

Reply:
That's stupid. We should fix that

Reply:
Or get rid of `self._current_class` completely. Is it needed?

Reply:
The use of `self._current_class` here:
https://github.com/pyccel/pyccel/blob/275b85717e8bfc2f42eab6646ec3007f3753aa09/pyccel/parser/semantic.py#L2102-L2104
seems especially dangerous. If we are going to rely on this variable for anything then we need to make sure it can't change at random. I would feel safer if we used the `cls_base` of the lhs of a `DottedName` to find the class as this information is not global

Reply:
Fixed.